qa/L0_e2e/test_model.py

```diff
@@ -0,0 +1,563 @@
+    def predict_networked(arr):
+        try:
+            return triton_predict(
+                model_name,
+                arr,
+                model_version=model_version,
+                protocol=protocol,
+                host=host,
+                port=[http_port, grpc_port][protocol == 'grpc'],
+                shared_mem=None,
+                predict_proba=predict_proba,
+                num_classes=num_classes,
+                attempts=retries
+            )
+        except Exception:
+            return (None, traceback.format_exc())
+
+    def predict_shared(arr):
+        try:
+            return triton_predict(
+                model_name,
+                arr,
+                model_version=model_version,
+                protocol=protocol,
+                host=host,
+                port=[http_port, grpc_port][protocol == 'grpc'],
+                shared_mem='cuda',
+                predict_proba=predict_proba,
+                num_classes=num_classes,
+                attempts=retries
+            )
+        except Exception:
+            return (None, traceback.format_exc())
+
+    queue = Queue()
+    results = []
+
+    def predict_worker():
+        while True:
+            try:
+                next_input = queue.get()
+            except Empty:
+                continue
+            if next_input is None:
+                queue.task_done()
+                return
+            arr, indices, sharing = next_input
+            if sharing is None:
+                results.append(
+                    (indices, predict_networked(arr))
+                )
+            elif sharing == 'cuda':
+                results.append(
+                    (indices, predict_shared(arr))
+                )
+            queue.task_done()
+
+    pool = [
+        threading.Thread(target=predict_worker) for _ in range(concurrency)
+    ]
+    for thread in pool:
+        thread.daemon = True
+        thread.start()
+
+    start = perf_counter()
+    for batch_size_ in batch_sizes:
+        for i in range(
+            total_rows // batch_size_ +
+            int(bool(total_rows % batch_size_))
+        ):
+            indices = (
+                i * batch_size_,
+                min((i + 1) * batch_size_, total_rows)
+            )
+
+            arr = total_batch[indices[0]: indices[1]]
+            queue.put((arr, indices, shared_mem[i % len(shared_mem)]))
+    for _ in range(concurrency):
+        queue.put(None)
+
+    for thread in pool:
+        thread.join(timeout=timeout)
+        if thread.is_alive():
+            raise RuntimeError("Test run exceeded timeout")
+    total = perf_counter() - start
+
+    throughput = len(batch_sizes) * total_rows / total
+    per_sample = 1 / throughput
+
+    request_latency = 0
+    row_latency = 0
+    for indices, (triton_result, batch_latency) in results:
+        if triton_result is None:
+            raise RuntimeError(
+                f'Prediction failed with error:\n\n{batch_latency}'
```

Comment:
Maybe add a comment indicating that, when an error occurs, `batch_latency` contains the error code?